File: cracking_the_coding_interview/8-13.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct Box {
    int wi, hi, di;
};

int findHighest(const vector<Box>& boxes){
    auto cmp = [](Box left, Box right) {
        if (left.hi != right.hi)
            return left.hi < right.hi;
        if (left.wi != right.wi)
            return left.wi < right.wi;
        return left.di < right.di;
    };
    priority_queue<Box, vector<Box>, decltype(cmp)> q(cmp);
    for(Box box: boxes)
        q.push(box);
    
    vector<pair<int, Box>> heights;
    heights.push_back(pair<int, Box>(0, Box{INT_MAX, INT_MAX, INT_MAX}));
    while(!q.empty()){
        Box box = q.top();
        q.pop();
        for(auto h: heights){
            if(h.second.wi > box.wi && h.second.di > box.di)
                heights.push_back(pair<int, Box>(h.first + box.hi, box));
        }
    }

    int ans = 0;
    for(auto h: heights){
        if(ans < h.first)
            ans = h.first;
    }
    return ans;
}

int main(void){
    Box box1 = {1, 1, 1};
    Box box2 = {1, 5, 1};
    Box box3 = {3, 2, 3};
    Box box4 = {3, 3, 2};
    Box box5 = {2, 3, 3};
    vector<Box> input = {box1, box2, box3, box4, box5};
    cout << findHighest(input) << endl;;
    return 0;
}

File: cracking_the_coding_interview/8-13.py
import heapq
import sys

class Box:
    def __init__(self, h, w, d):
        self.h = h
        self.w = w
        self.d = d


def findHighest(boxes):
    h = []
    for box in boxes:
        heapq.heappush(h, (-box.h, -box.w, -box.d))

    heights = [[0, (sys.maxsize, sys.maxsize, sys.maxsize)]]
    while len(h):
        box = heapq.heappop(h)
        for height in heights:
            if -box[0] < height[1][0] and -box[1] < height[1][1] and -box[2] < height[1][2]:
                heights.append([height[0] - box[0], (-box[0], -box[1], -box[2])])
        
    ans = 0
    for height in heights:
        if ans < height[0]:
            ans = height[0]
    return ans



if __name__ == '__main__':
    boxes = []
    boxes.append(Box(1, 1, 2))
    boxes.append(Box(3, 1, 2))
    boxes.append(Box(2, 2, 3))
    boxes.append(Box(3, 3, 3))
    print(findHighest(boxes))